Added standalone light client patch (#2130)
ethereum/eth2.0-specs: specs/lightclient/beacon-chain.md

    # Sync committee aggregate signature
    sync_committee_bits: Bitvector[SYNC_COMMITTEE_SIZE]
    sync_committee_signature: BLSSignature
```

#### `BeaconState`

```python
class BeaconState(phase0.BeaconState):
    # Sync committees
    current_sync_committee: SyncCommittee
    next_sync_committee: SyncCommittee
```

### New containers

#### `SyncCommittee`

```python
class SyncCommittee(Container):
    pubkeys: Vector[BLSPubkey, SYNC_COMMITTEE_SIZE]
    pubkey_aggregates: Vector[BLSPubkey, SYNC_COMMITTEE_SIZE // SYNC_COMMITTEE_PUBKEY_AGGREGATES_SIZE]
```

## Helper functions

### Beacon state accessors

#### `get_sync_committee_indices`

```python
def get_sync_committee_indices(state: BeaconState, epoch: Epoch) -> Sequence[ValidatorIndex]:
    """
    Return the sync committee indices for a given state and epoch.
    """
    base_epoch = Epoch((max(epoch // EPOCHS_PER_SYNC_COMMITTEE_PERIOD, 1) - 1) * EPOCHS_PER_SYNC_COMMITTEE_PERIOD)
    active_validator_indices = get_active_validator_indices(state, base_epoch)
    active_validator_count = uint64(len(active_validator_indices))
    seed = get_seed(state, base_epoch, DOMAIN_SYNC_COMMITTEE)
    i, sync_committee_indices = 0, []

>     while len(sync_committee_indices) < SYNC_COMMITTEE_SIZE:

Reviewer: > Is that intentional?

Yes, it's intentional! The reason is to more easily handle the edge case where there are less than `SYNC_COMMITTEE_SIZE` active validators.

        shuffled_index = compute_shuffled_index(uint64(i % active_validator_count), active_validator_count, seed)
        candidate_index = active_validator_indices[shuffled_index]
        random_byte = hash(seed + uint_to_bytes(uint64(i // 32)))[i % 32]
        effective_balance = state.validators[candidate_index].effective_balance

>         if effective_balance * MAX_RANDOM_BYTE >= MAX_EFFECTIVE_BALANCE * random_byte:

Reviewer 2: `MAX_RANDOM_BYTE` is not defined globally

It was previous defined within `compute_proposer_index`

            sync_committee_indices.append(candidate_index)
        i += 1
    return sync_committee_indices
```

#### `get_sync_committee`

```python
def get_sync_committee(state: BeaconState, epoch: Epoch) -> SyncCommittee:
    """
    Return the sync committee for a given state and epoch.
    """
    indices = get_sync_committee_indices(state, epoch)
    validators = [state.validators[index] for index in indices]
    pubkeys = [validator.pubkey for validator in validators]
    aggregates = [
        bls.AggregatePKs(pubkeys[i:i + SYNC_COMMITTEE_PUBKEY_AGGREGATES_SIZE])
        for i in range(0, len(pubkeys), SYNC_COMMITTEE_PUBKEY_AGGREGATES_SIZE)
    ]
    return SyncCommittee(pubkeys, aggregates)
```

### Block processing

```python
def process_block(state: BeaconState, block: BeaconBlock) -> None:
    phase0.process_block(state, block)
    process_sync_committee(state, block.body)
```

#### Sync committee processing

```python
def process_sync_committee(state: BeaconState, body: BeaconBlockBody) -> None:
    # Verify sync committee aggregate signature signing over the previous slot block root

>     previous_slot = max(state.slot, Slot(1)) - Slot(1)

Reviewer 3: Maybe a `get_previous_slot` accessor like `get_previous_epoch`

    committee_indices = get_sync_committee_indices(state, get_current_epoch(state))
    participant_indices = [committee_indices[i] for i in range(len(committee_indices)) if body.sync_committee_bits[i]]

>     participant_pubkeys = [state.validators[participant_index].pubkey for participant_index in participant_indices]

Reviewer 3: Might consider getting the pubkeys from the `SyncCommittee` object. Will make execution of this function with partials more straight forward for light clients.

```suggestion
    committee_pubkeys = state.current_sync_committee.pubkeys
    participant_pubkeys = [pubkey for pubkey, bit in zip(committee_pubkeys, body.sync_committee_bits) if bit]
```

    domain = get_domain(state, DOMAIN_SYNC_COMMITTEE, compute_epoch_at_slot(previous_slot))
    signing_root = compute_signing_root(get_block_root_at_slot(state, previous_slot), domain)
    assert bls.FastAggregateVerify(participant_pubkeys, signing_root, body.sync_committee_signature)

    # Reward sync committee participants
    participant_rewards = Gwei(0)

>     active_validator_count = uint64(len(get_active_validator_indices(state, get_current_epoch(state))))

Reviewer 2: Should we use `base_epoch` instead of `current_epoch`? 

It aligns with `get_sync_committee_indices`

>     for participant_index in participant_indices:

Reviewer 2: Should we exclude slashed validators for rewards?

        base_reward = get_base_reward(state, participant_index)

>         reward = Gwei(base_reward * active_validator_count // len(committee_indices) // SLOTS_PER_EPOCH)

Reviewer 3: Why is this reward so high?

So that on average you get a base_reward per epoch from sync committee duties?
@vbuterin

        increase_balance(state, participant_index, reward)
        participant_rewards += reward

    # Reward beacon proposer

>     increase_balance(state, get_beacon_proposer_index(state), Gwei(participant_rewards // PROPOSER_REWARD_QUOTIENT))

Reviewer 3: If we are going to do this like we do other proposer rewards, we need to temper the participant rewards by the `1 - PROPOSER_REWARD_QUOTIENT`
So that in aggregate a max single base_reward is given per participant.

something like the following:
```python
proposer_reward = Gwei(0)
for participant_index in participant_indices:
    base_reward = get_base_reward(state, participant_index)
    max_participant_reward = base_reward - base_reward // PROPOSER_REWARD_QUOTIENT
    reward = Gwei(max_participant_reward * active_validator_count // len(committee_indices) // SLOTS_PER_EPOCH)
    increase_balance(state, participant_index, reward)
    proposer_reward += base_reward // PROPOSER_REWARD_QUOTIENT
increase_balance(state, get_beacon_proposer_index(state), Gwei(proposer_reward))
```

>     increase_balance(state, get_beacon_proposer_index(state), Gwei(participant_rewards // PROPOSER_REWARD_QUOTIENT))

Reviewer 4: Should this new sync committee reward component be deducted from `get_inactivity_penalty_deltas`((https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/beacon-chain.md#components-of-attestation-deltas) `penalties` component for the well-performed validators? i.e., modify line `penalties[index] += Gwei(BASE_REWARDS_PER_EPOCH * base_reward - get_proposer_reward(state, index))`.

Reviewer 4: ^^^^ The above question was answered during the call, do not need to deduct the sync committee rewards.

Another question about light client reward + inactivity leak accounting:

- Since other reward components are all accounted during **epoch-processing**, they were calculated with (1) if `is_in_inactivity_leak(state)` is True, then pre-issue the `base_reward` and (2) cancel the pre-issued reward in `get_inactivity_penalty_deltas`.
- However, the light client committee rewards are calculated during the **block-processing**. So it cannot be detected with `is_in_inactivity_leak(state)`.
- It seems #2140 form can fix this issue well by using flags. Does `lightclient` patch require `accounting-reform` patch? Or there is a better way to delay the calculation for this case?

Reviewer 4: Update: applied @djrtwo 's suggestion in #2141 (https://github.com/ethereum/eth2.0-specs/pull/2141/commits/1b7b842aaff3c133266f67b2c0dd6811e580bf88)

        candidate_index = active_validator_indices[shuffled_index]
        random_byte = hash(seed + uint_to_bytes(uint64(i // 32)))[i % 32]
        effective_balance = state.validators[candidate_index].effective_balance
        if effective_balance * MAX_RANDOM_BYTE >= MAX_EFFECTIVE_BALANCE * random_byte:
            sync_committee_indices.append(candidate_index)
        i += 1
    return sync_committee_indices
```

#### `get_sync_committee`

```python
def get_sync_committee(state: BeaconState, epoch: Epoch) -> SyncCommittee:
    """
    Return the sync committee for a given state and epoch.
    """
    indices = get_sync_committee_indices(state, epoch)
    validators = [state.validators[index] for index in indices]
    pubkeys = [validator.pubkey for validator in validators]
    aggregates = [
        bls.AggregatePKs(pubkeys[i:i + SYNC_COMMITTEE_PUBKEY_AGGREGATES_SIZE])
        for i in range(0, len(pubkeys), SYNC_COMMITTEE_PUBKEY_AGGREGATES_SIZE)
    ]
    return SyncCommittee(pubkeys, aggregates)
```

### Block processing

```python
def process_block(state: BeaconState, block: BeaconBlock) -> None:

>     phase0.process_block(state, block)

Reviewer 4: Note that by `phase0.process_block`, it uses the phase 0 package.

1. The constants and the function calls `process_block` would be *phase 0 version*. It may cause some side effects that we may easily overlook.
2. You are passing `(lightclient_patch.state, lightclient_patch.block)` to `process_block(state: phase0.BeaconState, block: phase0.BeaconBlock)`.

Saving 3 lines probably is not worth it. 😅

```

#### `decompactify_validator`

```python
def decompactify_validator(compact_validator: CompactValidator) -> Tuple[ValidatorIndex, bool, Gwei]:
    """
    Return the validator index, slashed status and effective balance for a given compact validator.
    """
    index = ValidatorIndex(compact_validator >> 16)  # from bits 16-63
    slashed = bool((compact_validator >> 15) % 2)  # from bit 15
    effective_balance = Gwei(compact_validator & (2**15 - 1)) * EFFECTIVE_BALANCE_INCREMENT  # from bits 0-14
    return (index, slashed, effective_balance)
```

### Beacon state accessors

#### `get_sync_committee_indices`

```python
def get_sync_committee_indices(state: BeaconState, epoch: Epoch) -> Sequence[ValidatorIndex]:
    """
    Return the sync committee indices for a given state and epoch.
    """
    start_epoch = Epoch((max(epoch // EPOCHS_PER_SYNC_COMMITTEE_PERIOD, 1) - 1) * EPOCHS_PER_SYNC_COMMITTEE_PERIOD)
    active_validator_count = uint64(len(get_active_validator_indices(state, start_epoch)))
    sync_committee_size = min(active_validator_count, MAX_SYNC_COMMITTEE_SIZE)

>     seed = get_seed(state, base_epoch, DOMAIN_SYNC_COMMITTEE)

Reviewer 4: `start_epoch`? Fixed in https://github.com/ethereum/eth2.0-specs/pull/2133
```suggestion
    seed = get_seed(state, start_epoch, DOMAIN_SYNC_COMMITTEE)
```

Reviewer: I knew I must have introduced a bug somewhere 😂 

@hwwhww: Does everything else pass? For example, are all the types casted properly and the ToC properly built?

Reviewer: Used `base_epoch` everywhere!

    return [compute_shuffled_index(uint64(i), active_validator_count, seed) for i in range(sync_committee_size)]
```

### `get_sync_committee`

```python
def get_sync_committee(state: BeaconState, epoch: Epoch) -> SyncCommittee:
    """
    Return the sync committee for a given state and epoch.
    """
    indices = get_sync_committee_indices(state, epoch)
    validators = [state.validators[index] for index in indices]
    pubkeys = [validator.pubkey for validator in validators]
    compact_validators = [compactify_validator(i, v.slashed, v.effective_balance) for i, v in zip(indices, validators)]

>     return SyncCommittee(pubkeys, bls.AggregatePubkeys(pubkeys), compact_validators)

Reviewer 4: ```suggestion
    return SyncCommittee(pubkeys, bls.AggregatePKs(pubkeys), compact_validators)
```

fixed in https://github.com/ethereum/eth2.0-specs/pull/2133

```python
def get_light_client_committee(beacon_state: BeaconState, epoch: Epoch) -> Sequence[ValidatorIndex]:
    """
    Return the light client committee of no more than ``LIGHT_CLIENT_COMMITTEE_SIZE`` validators.
    """
    source_epoch = (max(epoch // LIGHT_CLIENT_COMMITTEE_PERIOD, 1) - 1) * LIGHT_CLIENT_COMMITTEE_PERIOD
    active_validator_indices = get_active_validator_indices(beacon_state, source_epoch)
    seed = get_seed(beacon_state, source_epoch, DOMAIN_LIGHT_CLIENT)
    return [
        compute_shuffled_index(i, active_validator_indices, seed)
        for i in range(min(active_validator_indices, LIGHT_CLIENT_COMMITTEE_SIZE))
    ]
```

### Block processing

```python
def process_block(state: BeaconState, block: BeaconBlock) -> None:
    phase0.process_block(state, block)
    process_light_client_signature(state, block.body)
```

#### Light client processing

```python
def process_light_client_signature(state: BeaconState, block_body: BeaconBlockBody) -> None:
    committee = get_light_client_committee(state, get_current_epoch(state))
    previous_slot = max(state.slot, 1) - 1
    previous_block_root = get_block_root_at_slot(state, previous_slot)

>     # Light clients sign over the previous block root

Reviewer 5: May be a heavy-handed edit but I'm trying to sort out what kind of security guarantees we get here (and how) so I don't think it hurts to be as precise as possible...

```suggestion
    # Light client committees sign over the previous block root
```

Reviewer 6: ,my name is Aura ,I saved your group page long time ago ,after years hope all is good with you .
Always believed in it so I had 0.5 Ethereum that I keep it for a long time for my daughter and lost it to a scammer. 
Now after I m putting the baby to sleep I m reading and watching videos. 
Is any chance to recover it?.
I have 200$ I know is not much and I know is late to invest ?
I know but I dont have much.

Thank youuu

    signing_root = compute_signing_root(
        previous_block_root,
        get_domain(state, DOMAIN_LIGHT_CLIENT, compute_epoch_at_slot(previous_slot))
    )
    
    participants = [
        committee[i] for i in range(len(committee)) if block_body.light_client_bits[i]
    ]
    
    signer_pubkeys = [
        state.validators[participant].pubkey for participant in participants
    ]
    
    assert bls.FastAggregateVerify(signer_pubkeys, signing_root, block_body.light_client_signature)
    
    # Process rewards
    total_reward = Gwei(0)
    active_validator_count = len(get_active_validator_indices(beacon_state, get_current_epoch(state)))
    for participant in participants:

>         reward = get_base_reward(state, participant) * active_validator_count // len(committee)

Reviewer 5: do we want to further reduce the reward by `SLOTS_PER_EPOCH`? e.g.

```python
reward = ...
reward = reward // SLOTS_PER_EPOCH
```

`BASE_REWARDS_PER_EPOCH` has been increased to `5` but as it stands this reward is assessed per slot.

Reviewer 5: and it *feels* like we also want to reduce the per-participant reward (proportionately) to account for the proposer reward... unless i'm missing something w/ the rewards rationale here.

    if is_in_inactivity_leak(state):
        matching_target_attestations = get_matching_target_attestations(state, get_previous_epoch(state))
        matching_target_attesting_indices = get_unslashed_attesting_indices(state, matching_target_attestations)
        for index in get_eligible_validator_indices(state):
            # Penalize validator so that optimal attestation performance is rewarded with one base reward per epoch
            base_reward = get_base_reward(state, index)
            penalties[index] += Gwei((BASE_REWARDS_PER_EPOCH - 1) * base_reward - get_proposer_reward(state, index))
            if index not in matching_target_attesting_indices:
                effective_balance = state.validators[index].effective_balance
                penalties[index] += Gwei(effective_balance * get_finality_delay(state) // INACTIVITY_PENALTY_QUOTIENT)

    # No rewards associated with inactivity penalties
    rewards = [Gwei(0) for _ in range(len(state.validators))]
    return rewards, penalties
```

#### Final updates

*Note*: The function `process_final_updates` is modified to handle sync committee updates.

```python
def process_final_updates(state: BeaconState) -> None:

>     phase0.process_final_updates(state)

Reviewer 4: Same as https://github.com/ethereum/eth2.0-specs/pull/2130#discussion_r537745449
Some notes:
- The current logic looks "fine" since the light client patch is irrelevant to the existing `phase0.process_final_updates`. 
- Still, it's dangerous to call the helper function in the `phase0` package directly.
- Note that if #2150 is included in the HF1, #2150 will unfold the full `process_final_updates` anyway there.

Maybe add a note here for now?
```suggestion
    # FIXME: unfold the full `process_final_updates` to avoid side effects.
    phase0.process_final_updates(state)
```
